pkg/services/ngalert/schedule/fetcher.go

```diff
@@ -3,15 +3,110 @@ package schedule
 import (
 	"context"
 	"fmt"
+	"sync"
 	"time"
 
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/ngalert/metrics"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 )
 
+// RulesStore is a store that provides alert rules for scheduling
+type RulesStore interface {
+	GetAlertRulesKeysForScheduling(ctx context.Context) ([]models.AlertRuleKeyWithVersion, error)
+	GetAlertRulesForScheduling(ctx context.Context, query *models.GetAlertRulesForSchedulingQuery) error
+}
+
+type FetcherCfg struct {
+	ReloadInterval       time.Duration
+	DisableGrafanaFolder bool
+}
+
+// BackgroundFetcher periodically refreshes rules and folders in the background.
+type BackgroundFetcher struct {
+	cfg       FetcherCfg
+	ticker    *time.Ticker
+	ruleStore RulesStore
+	metrics   *metrics.Scheduler
+	logger    log.Logger
+
+	mu                 sync.Mutex
+	latestRules        []*models.AlertRule
+	latestFolderTitles map[models.FolderKey]string
+}
+
+func NewBackgroundFetcher(cfg FetcherCfg, ruleStore RulesStore, metrics *metrics.Scheduler, logger log.Logger) *BackgroundFetcher {
+	return &BackgroundFetcher{
+		cfg:                cfg,
+		ticker:             time.NewTicker(cfg.ReloadInterval),
+		ruleStore:          ruleStore,
+		metrics:            metrics,
+		logger:             logger,
+		latestRules:        make([]*models.AlertRule, 0),
+		latestFolderTitles: make(map[models.FolderKey]string),
+	}
+}
+
+// Refresh forces a synchronous refresh and blocks until the refresh is done.
+// This is thread-safe with the periodic background refresh.
+func (f *BackgroundFetcher) Refresh(ctx context.Context) {
+	f.updateSchedulableAlertRules(context.Background())
+}
+
+func (f *BackgroundFetcher) Run(ctx context.Context) error {
+	f.logger.Info("Starting rules fetcher")
+	for {
+		select {
+		case <-f.ticker.C:
+			f.logger.Debug("Refreshing rules from storage")
+			f.updateSchedulableAlertRules(context.Background())
+		case <-ctx.Done():
+			f.ticker.Stop()
+			f.logger.Info("Fetcher is shut down")
+			return nil
+		}
+	}
+}
+
+func (f *BackgroundFetcher) Rules() ([]*models.AlertRule, map[models.FolderKey]string) {
+	return f.latestRules, f.latestFolderTitles
+}
+
+func (f *BackgroundFetcher) updateSchedulableAlertRules(ctx context.Context) {
+	start := time.Now()
+	defer func() {
+		f.metrics.UpdateSchedulableAlertRulesDuration.Observe(
+			time.Since(start).Seconds(),
+		)
+	}()
+
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	q := models.GetAlertRulesForSchedulingQuery{
+		PopulateFolders: !f.cfg.DisableGrafanaFolder,
+	}
+	if err := f.ruleStore.GetAlertRulesForScheduling(ctx, &q); err != nil {
```

Comment:
The `GetAlertRulesKeysForScheduling` optimization is notably missing here. This is to support the (future) reloading based on fingerprint rather than version. The optimization is incompatible with this, as it queries the version field to shortcut fetching the full rule definitions.

By making this async, we have bought more time for this query to happen. The optimization makes the p99 for fetching slower while making the average fetch faster.

If we decide we want to put the optimization back, we can still do so in the future by putting fingerprint into the database.

Reply:
I think this is fine. Have you done any experiments with different size databases to see how long it takes? For example, 50 rules, 500 rules, 5000 rules, 50,000 rules?

Reply:
Yes, I actually started this process by creating a bench test: https://github.com/grafana/grafana/compare/main...alexweav/bench-test-fetcher-shortcut

On a few thousand rules with a local database (no network) the optimization was only saving us less than 50ms in the case where rules AREN'T reloaded. I'd expect network to make it significantly worse, I haven't yet run the bench test against a remote database.

I didn't test the remote case because I realized we have a metric here... it was easier to just see what the existing instances were doing directly.

The optimization has the opposite effect in the case where a reload happens, because it has to do two queries. Of course, 99% of queries do not have a reload. The benchmark is effectively trading faster average case for worse p99 case.